From e468b416d6d19c87e65dbbee08849cd1f648ae13 Mon Sep 17 00:00:00 2001
From: Tong Tiangen <tongtiangen@huawei.com>
Date: Thu, 12 Jan 2023 11:22:01 +0000
Subject: [PATCH 002/905] [Huawei] kasan: add kasan support for memcpy_mcs()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188249

--------------------------------

The function memcpy_mcs() is designed to copy in memory and should be
supported by kasan.

Signed-off-by: Tong Tiangen <tongtiangen@huawei.com>
---
 mm/kasan/common.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/mm/kasan/common.c b/mm/kasan/common.c
index b48b62673b0f..ac2ec0befe43 100644
--- a/mm/kasan/common.c
+++ b/mm/kasan/common.c
@@ -109,6 +109,18 @@ void *memcpy(void *dest, const void *src, size_t len)
 	return __memcpy(dest, src, len);
 }
 
+#ifdef __HAVE_ARCH_MEMCPY_MC
+#undef memcpy_mcs
+unsigned long memcpy_mcs(void *dest, const void *src, size_t len)
+{
+	if (!check_memory_region((unsigned long)src, len, false, _RET_IP_) ||
+	    !check_memory_region((unsigned long)dest, len, true, _RET_IP_))
+		return (unsigned long)len;
+
+	return __memcpy_mcs(dest, src, len);
+}
+#endif
+
 /*
  * Poisons the shadow memory for 'size' bytes starting from 'addr'.
  * Memory addresses should be aligned to KASAN_SHADOW_SCALE_SIZE.
-- 
2.17.1

